### PR TITLE
add audio jitter simulator testing tool

### PR DESCRIPTION
**Pull request by zzmp**

### CODE REVIEW

**Comment by zzmp** [comment]
QA:

- Smoke test audio
- There are two new settings, only available in DEV/PR builds, both under `Settings > Audio`
  - `Packet jitter simulator` - overrides `Packet throttle threshold` - audio packets are held, and there is a 60% chance on any audio packet receipt that the held packets will be flushed. This simulates varied network simulation (jitter) because our internet is too good :/
  - `Packet throttle threshold` - holds the set number of audio packets before flushing them, forcing the jitter buffer to grow to be the set number of frames

If you use these settings, your jitter should adjust to compensate, meaning you will have more latency in your audio.
You are not testing that the jitter buffer works as expected, you are testing that the two new settings perturb the audio packets as expected. So if your audio gets worse, it is probably still a QA pass.

**Comment by hifi-gustavo** [comment]
Build console output for commit: f6403f78220661426687f3542189177d54efa6ad available at: https://gist.github.com/00335ba440a4042db92ffb9ee11997ae<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8610-f6403f7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8610-f6403f7.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 53f5899a69493cd8ac4d04048fc0da733908a0bb available at: https://gist.github.com/00335ba440a4042db92ffb9ee11997ae<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8610-53f5899.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8610-53f5899.dmg

**Comment by kevinmthomasse** [comment]
testing

**Comment by kevinmthomasse** [comment]
## QA Test
**QA Tester**: KevinMThomas
**QA Pass**: FAILURE
**Additional Test Results + Observations**: Tested in ```dev-demo``` and selected under ```Settings > Audio
Packet jitter simulator``` - I experience absolutely no loss of audio quality of any kind whatsoever.  I then tested ```packet throttle threshold``` brought the dial up to 160 and tested to which again there was absolutely no loss of audio quality.  The only issue when exiting interface it did a hard white-screen crash to which I had to ctrl-alt-delete.  Log file attached for deeper inspection.
[hifi-log.txt](https://github.com/highfidelity/hifi/files/475809/hifi-log.txt)

**Comment by kencooke** [comment]
Just want to point out, this simulates packet loss, not jitter. That will give misleading results if used for jitter-buffer testing.

**Comment by zzmp** [comment]
@kencooke it eventually delivers all packets, it just holds them and delivers them all at once.
Why are you saying it is packet loss and not jitter?

**Comment by kencooke** [comment]
@zzmp Sorry, I didn't see how this code:
``if (randFloat() < 0.6f) {
      flush(); // 60% of the time, it works every time
 }``
could do anything other than drop packets 60% of the time. Perhaps a comment could make it more clear. What is the effective distribution of the jitter?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 80bce953f74a43c998402c197b51592dc45d9fe6 available at: https://gist.github.com/00335ba440a4042db92ffb9ee11997ae<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8610-80bce95.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8610-80bce95.dmg

File: libraries/audio-client/src/AudioClient.cpp [hunk 1]
```diff
@@ -551,31 +548,53 @@ void AudioClient::handleAudioDataPacket(QSharedPointer<ReceivedMessage> message)
...
-AudioClient::Gate::Gate(AudioClient* audioClient, int threshold) :
-    _audioClient(audioClient),
-    _threshold(threshold) {}
+AudioClient::Gate::Gate(AudioClient* audioClient) :
+    _audioClient(audioClient) {}
+
+void AudioClient::Gate::setIsSimulatingJitter(bool enable) {
+    std::lock_guard<std::mutex> lock(_mutex);
+    flush();
+    _isSimulatingJitter = true;
```
Comment (zzmp) [inline libraries/audio-client/src/AudioClient.cpp]: oops

**Comment by zzmp** [comment]
@kevinmthomasse try testing by listening to this:
```
// given me by @kencooke
var sound = SoundCache.getSound("http://xiph.org/~xiphmont/demo/1kHz-0-dither.wav");
Audio.playSound(sound, { position: Vec3.sum(MyAvatar.position, Quat.getFront(MyAvatar.orientation)), loop: true, localOnly: false, volume: 1.0 });
```

You should be able to hear breaks every 5-10s from the jitter simulator.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 45b306fc97054774b7f451c56e8cb04320d5e4c9 available at: https://gist.github.com/00335ba440a4042db92ffb9ee11997ae<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8610-45b306f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8610-45b306f.dmg

**Comment by kevinmthomasse** [comment]
@zzmp asked to retest Settings > Audio Packet jitter simulator - I experience absolutely no loss of audio quality of any kind whatsoever.  QA PASS

**Comment by zzmp** [comment]
@kevinmthomasse you should have experienced *some* when the simulator was on.

**Comment by kevinmthomasse** [comment]
@zzmp I did not.  I set ```echo server audio``` on and listened carefully.  Since we are getting varied results between us there must be something in my setup that is different.

**Comment by zzmp** [comment]
Tested with @kevinmthomasse and it does vary the jitter.
On my machine it was causing starve a few times a minute, on his machine it rarely causes starve, but it still varies the jitter.

QA Pass. If I need to use this tool more I will make it vary the jitter more reliably.

**Comment by kencooke** [comment]
I installed and can confirm that with "packet jitter simulator" enabled, the max inter-packet timegaps (in Audio Network Stats) does increase. It has very little effect on the final audio output in my case, but I think that is expected.

**Pull request merged**